compiler/rustc_const_eval/src/check_consts/qualifs.rs

```diff
@@ -192,16 +183,32 @@ impl Qualif for NeedsNonConstDrop {
             return false;
         }
 
-        // FIXME(const_trait_impl): Reimplement const drop checking.
-        NeedsDrop::in_any_value_of_ty(cx, ty)
+        if cx.tcx.features().const_trait_impl() {
+            let destruct_def_id = cx.tcx.require_lang_item(LangItem::Destruct, Some(cx.body.span));
+            let infcx = cx.tcx.infer_ctxt().build(TypingMode::from_param_env(cx.param_env));
+            let ocx = ObligationCtxt::new(&infcx);
+            ocx.register_obligation(Obligation::new(
+                cx.tcx,
+                ObligationCause::misc(cx.body.span, cx.def_id()),
+                cx.param_env,
+                ty::Binder::dummy(ty::TraitRef::new(cx.tcx, destruct_def_id, [ty]))
+                    .to_host_effect_clause(cx.tcx, match cx.const_kind() {
+                        rustc_hir::ConstContext::ConstFn => ty::BoundConstness::Maybe,
+                        rustc_hir::ConstContext::Static(_)
+                        | rustc_hir::ConstContext::Const { .. } => ty::BoundConstness::Const,
+                    }),
+            ));
+            !ocx.select_all_or_error().is_empty()
+        } else {
+            NeedsDrop::in_any_value_of_ty(cx, ty)
+        }
     }
 
-    fn in_adt_inherently<'tcx>(
-        cx: &ConstCx<'_, 'tcx>,
-        adt: AdtDef<'tcx>,
-        _: GenericArgsRef<'tcx>,
-    ) -> bool {
-        adt.has_non_const_dtor(cx.tcx)
+    fn is_non_structural<'tcx>(cx: &ConstCx<'_, 'tcx>, adt: AdtDef<'tcx>) -> bool {
+        // Even a `const` dtor may have `~const` bounds that may need to
+        // be satisfied, so this becomes non-structural as soon as the
+        // ADT gets a destructor at all.
```

Comment:
Isn't the actual problem that the dtor may change the value and hence we can't recurse into the currently visible value? That's what the example you gave me said, anyway.

Reply:
Well, yes? And the fact that we run the dtor at all, which may execute arbitrary side-effects. I think the whole point is that the dtor simply stops being (value-wise) structural as soon as we have a dtor.

Reply:
I find the comment in the code confusing. I don't understand what this has to do with the dtor itself having bounds. Your example didn't involve anything being dropped in the dtor, it involved the dtor changing the state so that afterwards things get dropped that didn't exist in the original value.

Reply:
My example was to demonstrate why the logic you originally proposed is wrong. But it's only half of what "could go wrong" -- the comment also is attempting to justify why we don't also just check this:

* check if the type's Drop impl is const. If yes, we can propagate structurally, so return false.

Since that is really just redundantly replicating part of the check we do in `in_any_value`.

Reply:
> why we don't also just check this:

I find a comment saying "the drop impl can change the value being dropped, making value-based structural reasoning invalid" a much easier to understand, more compelling argument to explain this.

I still haven't understood the comment you put in the code.

Reply:
You're overlooking the fact that even if we were to, for example, eagerly fetch the `mir_for_ctfe` for the `Drop` and look through the `fn drop`'s body to see if it modifies the receiver struct at all, in an attempt to refine this qualification to be less false-positive (I'm not saying we should -- this is just for the sake of argument), then that check would *still* be unsufficient, since we would still need to make sure that the *conditionally constness* of the drop impl is actually satisfied.

Reply:
The fact that we simply check that a drop impl exists at all is deferring two kinds of checks that would make us execute non-const code in a const context:
1. A drop impl that may make calls out to other things that may or may not be const
2. A drop impl that modifies the value being dropped

Reply:
> A drop impl that may make calls out to other things that may or may not be const

Can't we rule that out since the impl has itself been const-checked already?

A drop impl cannot have bounds that aren't present on the type, after all. Though I actually have no idea how that interacts with `~const`.

Reply:
> Can't we rule that out since the impl has itself been const-checked already? A drop impl cannot have bounds that aren't present on the type, after all. Though I actually have no idea how that interacts with ~const.

No -- a drop impl may have additional `~const` conditions on top. Without this, we severely hinder the expressiveness of const drop, to the point that I have no idea what it would even be used for. 

For example, given some struct like:

```
struct DropAndCall<F: Fn()>(F);

impl<F> const Drop for DropAndCall
where
    F: ~const Fn(),
{
    fn drop(&mut self) {
        (self.0)(); // This should be allowed.
    }
}
```

We want to be able to drop `DropAndCall<F>` in const contexts only if `F` is `~const Fn()`. But it should be non-const droppable everywhere, which is guaranteed by the current drop impl compatibility check.

Reply:
>  a drop impl may have additional ~const conditions on top. 

Oh, that is the part I am missing. It is not how drop usually works so naturally I assume it's not how const drop works either.

> We want to be able to drop DropAndCall<F> in const contexts only if F is ~const Fn(). But it should be non-const droppable everywhere, which is guaranteed by the current drop impl compatibility check.

I guess the point is if we declared this type `struct DropAndCall<F: ~const Fn()>(F);` (assuming that's a thing), even non-dropping uses in const would be forbidden?

Has there ever been an RFC for this? If yes, I haven't seen it. In whatever becomes the tracking issue for const drop, this needs to be motivated and explained in a lot of detail, to make sure t-lang is aware what is happening.

Reply:
No, there was a closed RFC but it presumably didn't not actually plan for or specify the exact details and wrinkles. I can put more details in the tracking issue for https://github.com/rust-lang/rust/issues/133214.